Implement track (state) removal functionality

mwoehlke-kitware opened the pull request:
This implements low-level mechanisms (i.e. in `AnnotationGeometryContainer`) for removing geometries and entire tracks. These are not yet in use. It does not necessarily need to be merged anywhere in its current state, but I am pushing it now because the relevant code has been subject to frequent and heavy modification lately.

mwoehlke-kitware committed:
Factor out tracks

Rather than keeping several different member variables to track various
information related to tracks, factor out this logic into a new track
class, and keep a single map of tracks.

mwoehlke-kitware committed:
Fix off-by-one bug assigning geometry ID

matthewma7 commented:
Hi @mwoehlke-kitware, before you get further dependant on these changes I would like to frankly say that I don't think it's time for the refactoring in these changes. Could we work on new features and apply only necessary changes to existing code. If you think some refactoring is necessary let's discuss it before making the changes. I see the reason you might want to have some refactoring, but if the current organization works fine for the new features, I would say let's keep it intact. 
I'll leave some comment on the code inline.

matthewma7 commented on web_external/util/annotationGeometryParser.js:
```diff
@@ -1,154 +1,261 @@
+//=============================================================================
```
Could you tell me why you added these separators lines?

mwoehlke-kitware replied:
...because it's *much* easier to skim the code and identify critical logic boundaries (e.g. beginnings of functions) vs. having to read a wall of text. These are SOP in ViViA and fairly common in other vision projects (e.g. KWIVER).

matthewma7 replied:
Are the other projects JavaScript project? This basically add a line before all class and member without emphasis. I really don't see the different between adding this and a spare line. The class and member function are already identified by their intentation. And they are causing different eslinting errors.

mwoehlke-kitware replied:
They are *much easier* to pick out of the clutter compared to pure whitespace.

p.s. What does "without emphasis" mean?

matthewma7 commented on web_external/util/annotationGeometryParser.js:
```diff
@@ -1,154 +1,261 @@
+
+        // Create frame if needed
+        if (!this._frameMap.has(geometry.ts0)) {
+            this._frameMap.set(geometry.ts0, new Map());
```
I think the map of map is harder to understand than it was before with a flat organization.

mwoehlke-kitware replied:
...but it's *much* easier to *use*. It avoids several instances that before would have had to scan the whole array looking for a specific state. Now, I can just use key-lookup to directly access the desired value.

For example:
```javascript
    var geomToChange = Array.from(geomSet).find((geom) => {
        return geom.id1 === trackId && geom.ts0 === frame;
    });
```

...becomes:
```javascript
    var geomToChange = geomSet.get(trackId);
```

(...except that particular code wound up using the state map instead, but you get the idea.)

It may be (slightly) harder to understand up front, but it makes it *easier* to understand the code in many places the map is actually used.

matthewma7 commented on web_external/util/annotationGeometryParser.js:
```diff
@@ -1,154 +1,261 @@
-                this._editedGeom.add(geomToChange);
+        // Look up ID of possibly existing geometry
+        let state = {id1: trackId, ts0: frame};
+        let id0 = this._stateMap.get(state);
```
You probably still working on this but I would simply point out that this _stateMap is not defined at this moment.

mwoehlke-kitware replied:
Oops, yes... I was going back and forth between `_stateMap` and `_geomMap`, and some spots still had `_stateMap`. Fixed.

matthewma7 commented on web_external/util/annotationGeometryParser.js:
```diff
@@ -1,154 +1,261 @@
-                id0: this._id0++,
+            // Entirely new geometry; add it
+            let newGeom = new AnnotationGeometry({
+                id0: ++this._id0,
```
This is a good catch.

matthewma7 commented on web_external/util/annotationGeometryParser.js:
```diff
@@ -1,154 +1,261 @@
+//=============================================================================
+class AnnotationGeometryTrack {
```
Overall, this is what I think is unnecessary to refactor, I think a flat design is easier to understand.

mwoehlke-kitware replied:
I have to explicitly disagree. The flat design requires that every place that creates a track (there are at least two) needs to update multiple members. Refactoring moves these operations to the track constructor, which is a much more object-oriented design.

matthewma7 replied:
I understand your point. I think I need to see when the code in action.

mwoehlke-kitware replied:
Compare:

```javascript
    this._trackIds.add(trackId);
    this._trackRanges.set(trackId, [0, 0]);
    this._enableState.set(trackId, true);
```

...versus:

```javascript
    this._trackMap.set(trackId, new AnnotationGeometryTrack());
```

...and similar for removing a track.

Lookup-wise is a bit more of a wash, but I don't think post-refactor is any *less* logical. Instead of having a map per possible attribute of a track, i.e. `this._someTrackProp.get(id)`, you look up the property *on the track*, i.e. `this._tracks.get(id).someProp`. (To *me*,  this feels more natural, but YMMV.)

mwoehlke-kitware commented on web_external/util/annotationGeometryParser.js:
```diff
@@ -1,154 +1,261 @@
+                let state = {id1: trackId, ts0};
+                this._remove(frame, trackId, this._stateMap.get(state));
+            }
         }
```
...you're also supposed to tell me I forgot to remove the actual *track*, here :wink:. (Which is yet another time refactoring tracks into their own type is useful; I only need to change one member, not three.)

mwoehlke-kitware committed:
Fix frame range

Change the 'empty' frame range back to [MAX, MIN] (albeit using integers
this time), since this is needed for expanding an empty range to work
properly. This reportedly causes issues on the front-end, but having our
"empty" state be distinguishable from a valid state is probably a Good
Thing, and so we should probably fix the front-end to cope. (Immediately
prior to this change, the frame range would get "stuck" always having 0
as the lower bound.)

matthewma7 reviewed (approved):
I quickly verified with existing functionality.

mwoehlke-kitware committed:
Refactor geometry management

Refactor how AnnotationGeometryContainer tracks geometry:

- The primary data storage is now a map of frames to a map of tracks to
  geometries, rather than just a map of frames to an array of
  geometries. This makes it much easier to find a geometry to update,
  and incidentally makes renumbering tracks easier as well.

- The sets of added/edited geometry are now maps of geometry identifier
  to geometry, and there is now a map of state (frame plus track) to
  geometry identifier. This will make it possible to manage these lists
  once we start introducing the ability to remove geometry.

mwoehlke-kitware committed:
Implement (low level) removal

Add methods to AnnotationGeometryContainer to allow removing a geometry,
or an entire track, and to obtain the set of removals.

mwoehlke-kitware committed:
Implement removal of track states

Hook up handling of 'del' to remove the geometry of the track being
edited at the current frame. Adjust data-commit code to also commit
deletions.

mwoehlke-kitware commented:
Removal of states is now implemented and seems to be working. Removal of entire tracks is pending (partly due to interaction with activities). Note that I had to refactor the low-level code because Javascript can't use objects as `Map` keys.

matthewma7 reviewed (approved):
The latest state LGTM. Approving.
